src/OrleansRuntime/GrainDirectory/GrainDirectoryPartition.cs

```diff
@@ -289,17 +292,19 @@ internal virtual AddressAndTag AddSingleActivation(GrainId grain, ActivationId a
 
             if (!IsValidSilo(silo))
                 return result;
-            
+
+            IGrainInfo grainInfo;
             lock (lockable)
             {
-                if (!partitionData.ContainsKey(grain))
+                if (!partitionData.TryGetValue(grain, out grainInfo))
                 {
-                    partitionData[grain] = new GrainInfo();
+                    partitionData[grain] = grainInfo = new GrainInfo();
                 }
-                var grainInfo = partitionData[grain];
-                result.Address = grainInfo.AddSingleActivation(grain, activation, silo, registrationStatus);
-                result.VersionTag = grainInfo.VersionTag;
             }
+
+            result.Address = grainInfo.AddSingleActivation(grain, activation, silo, registrationStatus);
```

Comment:
Previously it was under the lock, if necessary it can be guarded with more fine-grained one.